myia/anf_ir_utils.py

```diff
@@ -17,3 +17,25 @@ def dfs(root: ANFNode, follow_graph: bool = False) -> Iterable[ANFNode]:
                 to_visit.append(in_)
         if isinstance(node.value, Graph) and follow_graph:
             to_visit.append(node.value.return_)
+
+
+def toposort(graph: Graph) -> Iterator[ANFNode]:
```

Comment:
I would need a more generic `toposort` which takes a `succ` function, i.e.

```python
T = TypeVar('T')
def toposort(root: T, succ: Callable[[T], Iterable[T]]) -> Iterable[T]:
```

I think it would also be useful to raise an exception if there is a cycle, instead of looping infinitely. That would help debug optimizations that accidentally introduce cycles, for example.